File: scripts/list-workspaces.py
#!/usr/bin/env python3
"""
Offline workspace audit tool for Camporee Conductor.
Reads conductor.db and workspace dirs directly — no server or Docker required.

Usage:
    python scripts/list-workspaces.py
    python scripts/list-workspaces.py --json
    python scripts/list-workspaces.py --user j2willey@gmail.com

Path resolution order (highest priority first):
    1. CLI flags:        --db / --workspaces
    2. Env vars:         CONDUCTOR_DB_PATH / WORKSPACE_PATH  (same vars used by the Node app)
    3. .env file:        loaded automatically from repo root if present
    4. Repo-relative defaults: ./data/shared/conductor.db  /  ./data/composer/workspaces

Recommended setup — add to ~/.bashrc on each machine:
    # Dev (nothing needed — repo-relative defaults work out of the box)
    # VPS:
    export CONDUCTOR_DB_PATH=/opt/camporee-conductor-data/shared/conductor.db
    export WORKSPACE_PATH=/opt/camporee-conductor-data/composer/workspaces

Then the alias is identical on both machines:
    alias cc-workspaces='python3 ~/ws/camporee-conductor/scripts/list-workspaces.py'
"""

import argparse
import json
import os
import sqlite3
import sys
from datetime import datetime, timezone
from pathlib import Path

# Resolve repo root
SCRIPT_DIR = Path(__file__).resolve().parent
REPO_ROOT  = SCRIPT_DIR.parent

# Load .env from repo root if present — populates os.environ
_env_file = REPO_ROOT / ".env"
if _env_file.exists():
    for _line in _env_file.read_text().splitlines():
        _line = _line.strip()
        if _line and not _line.startswith("#") and "=" in _line:
            _k, _, _v = _line.partition("=")
            os.environ.setdefault(_k.strip(), _v.strip().strip('"').strip("'"))

DEFAULT_DB = Path(os.environ.get(
    "CONDUCTOR_DB_PATH",
    str(REPO_ROOT / "data" / "shared" / "conductor.db")
))
DEFAULT_WORKSPACES = Path(os.environ.get(
    "WORKSPACE_PATH",
    str(REPO_ROOT / "data" / "composer" / "workspaces")
))

# Args
parser = argparse.ArgumentParser(description="Audit Camporee Conductor workspaces.")
parser.add_argument("--db",         default=str(DEFAULT_DB),         help="Path to conductor.db")
parser.add_argument("--workspaces", default=str(DEFAULT_WORKSPACES), help="Path to workspaces dir")
parser.add_argument("--user",       default=None,  help="Filter by email or user_id substring")
parser.add_argument("--json",       action="store_true",             help="Output JSON")
args = parser.parse_args()

db_path = Path(args.db)
ws_path = Path(args.workspaces)

if not db_path.exists():
    print(f"ERROR: DB not found at {db_path}", file=sys.stderr)
    print(f"       Set CONDUCTOR_DB_PATH env var or use --db flag.", file=sys.stderr)
    sys.exit(1)

# Read DB
con = sqlite3.connect(f"file:{db_path}?mode=ro", uri=True)
con.row_factory = sqlite3.Row
cur = con.cursor()
cur.execute("""
    SELECT ep.event_id, ep.role, ep.granted_at,
           up.user_id, up.display_name, up.email, up.is_sysadmin
    FROM event_permissions ep
    LEFT JOIN user_profiles up ON ep.user_id = up.user_id
    ORDER BY ep.granted_at ASC
""")
rows = cur.fetchall()
con.close()

def read_camporee_meta(event_id):
    p = ws_path / event_id / "camporee.json"
    if not p.exists():
        return {"title": "(missing camporee.json)", "year": "", "theme": "", "games": "?", "dir_exists": False}
    try:
        data = json.loads(p.read_text(encoding="utf-8"))
        meta = data.get("meta", {})
        return {
            "title": meta.get("title", "(no title)"),
            "year":  str(meta.get("year", "")),
            "theme": meta.get("theme", ""),
            "games": len(data.get("games", [])),
            "dir_exists": True,
        }
    except Exception as e:
        return {"title": f"(corrupt: {e})", "year": "", "theme": "", "games": "?", "dir_exists": True}

# Find orphan dirs
known_ids = {r["event_id"] for r in rows}
orphan_dirs = []
if ws_path.exists():
    for entry in sorted(ws_path.iterdir()):
        if entry.is_dir() and entry.name not in known_ids:
            orphan_dirs.append(entry.name)

# Group by event_id
events = {}
for r in rows:
    eid = r["event_id"]
    if eid not in events:
        events[eid] = {"event_id": eid, "permissions": [], **read_camporee_meta(eid)}
    granted_ts  = r["granted_at"]
    granted_str = (
        datetime.fromtimestamp(granted_ts, tz=timezone.utc).strftime("%Y-%m-%d")
        if granted_ts else ""
    )
    events[eid]["permissions"].append({
        "role":         r["role"],
        "user_id":      r["user_id"] or "",
        "display_name": r["display_name"] or "(no profile)",
        "email":        r["email"] or "(unknown)",
        "is_sysadmin":  bool(r["is_sysadmin"]),
        "granted_at":   granted_str,
    })

event_list = list(events.values())

# User filter
if args.user:
    f = args.user.lower()
    event_list = [
        e for e in event_list
        if any(
            f in (p["email"] or "").lower() or f in (p["user_id"] or "").lower()
            for p in e["permissions"]
        )
    ]

# JSON output
if args.json:
    out = {
        "generated_at": datetime.now(tz=timezone.utc).isoformat(),
        "db":           str(db_path),
        "workspaces":   str(ws_path),
        "event_count":  len(event_list),
        "orphan_count": len(orphan_dirs),
        "events":       event_list,
        "orphan_dirs":  [{"dir": d, **read_camporee_meta(d)} for d in orphan_dirs],
    }
    print(json.dumps(out, indent=2))
    sys.exit(0)

# Table output
def trunc(s, n):
    s = str(s or "")
    return (s[:n-1] + "~") if len(s) > n else s.ljust(n)

C = dict(event_id=14, title=30, year=6, games=5, role=8,
         email=30, display_name=22, granted=12, dir=4)

def row(*cells):
    keys = list(C.keys())
    return " | ".join(trunc(cells[i], C[keys[i]]) for i in range(len(cells)))

def divider():
    return "-+-".join("-" * w for w in C.values())

print(f"\nWorkspace audit -- {datetime.now().strftime('%Y-%m-%d %H:%M')}")
print(f"DB:         {db_path}")
print(f"Workspaces: {ws_path}")
if args.user:
    print(f"Filter:     {args.user}")
print(f"Events:     {len(event_list)}  |  Orphan dirs: {len(orphan_dirs)}\n")

print(row("event_id","title","year","games","role","email","display_name","granted","dir"))
print(divider())

for e in event_list:
    for p in e["permissions"]:
        print(row(
            e["event_id"], e["title"], e["year"], e["games"],
            p["role"], p["email"], p["display_name"], p["granted_at"],
            "Y" if e["dir_exists"] else "N",
        ))

if orphan_dirs:
    print("\n-- Orphan workspace dirs (no DB entry) --")
    for d in orphan_dirs:
        meta = read_camporee_meta(d)
        print(f"  {d}  ->  {meta['title']} {meta['year']}")

print()

File: scripts/list-workspaces.js
#!/usr/bin/env node
/**
 * Offline workspace audit tool.
 * Reads conductor.db and workspace directories directly — no server required.
 *
 * Usage:
 *   node scripts/list-workspaces.js
 *   node scripts/list-workspaces.js --json
 *   node scripts/list-workspaces.js --user <email_or_user_id>
 *
 * Env vars (optional, falls back to defaults):
 *   CONDUCTOR_DB_PATH   path to conductor.db
 *   WORKSPACE_PATH      path to composer workspaces dir
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { openConductorDb } from '../src/db/migrate.js';
import dotenv from 'dotenv';

dotenv.config();

const __filename = fileURLToPath(import.meta.url);
const __dirname  = path.dirname(__filename);
const PROJECT_ROOT = path.join(__dirname, '..');

const WORKSPACE_PATH = process.env.WORKSPACE_PATH
    || path.join(PROJECT_ROOT, 'data', 'composer', 'workspaces');

// --- CLI args ---
const args = process.argv.slice(2);
const jsonMode  = args.includes('--json');
const userIdx   = args.indexOf('--user');
const userFilter = userIdx !== -1 ? args[userIdx + 1] : null;

// --- Open DB ---
const db = openConductorDb();

// --- Query: all owner rows joined with user profiles ---
const rows = db.prepare(`
    SELECT
        ep.event_id,
        ep.role,
        up.user_id,
        up.display_name,
        up.email,
        up.is_sysadmin,
        ep.granted_at
    FROM event_permissions ep
    LEFT JOIN user_profiles up ON ep.user_id = up.user_id
    ORDER BY ep.granted_at ASC
`).all();

// --- Build camporee map: event_id → { title, year, theme, games } ---
function readCamporeeTitle(eventId) {
    const p = path.join(WORKSPACE_PATH, eventId, 'camporee.json');
    if (!fs.existsSync(p)) return { title: '(missing camporee.json)', year: '', theme: '' };
    try {
        const data = JSON.parse(fs.readFileSync(p, 'utf8'));
        const gameCount = data.games?.length ?? 0;
        return {
            title:  data.meta?.title  || '(no title)',
            year:   data.meta?.year   || '',
            theme:  data.meta?.theme  || '',
            games:  gameCount,
        };
    } catch {
        return { title: '(corrupt camporee.json)', year: '', theme: '', games: '?' };
    }
}

// --- Also list workspace dirs that have no DB entry (orphans) ---
let orphanDirs = [];
if (fs.existsSync(WORKSPACE_PATH)) {
    const allDirs = fs.readdirSync(WORKSPACE_PATH, { withFileTypes: true })
        .filter(e => e.isDirectory())
        .map(e => e.name);
    const knownIds = new Set(rows.map(r => r.event_id));
    orphanDirs = allDirs.filter(d => !knownIds.has(d));
}

// --- Group by event_id, collect all permission rows per event ---
const events = {};
for (const row of rows) {
    if (!events[row.event_id]) {
        events[row.event_id] = { event_id: row.event_id, permissions: [] };
    }
    events[row.event_id].permissions.push({
        role:         row.role,
        user_id:      row.user_id,
        display_name: row.display_name || '(no profile)',
        email:        row.email        || '(unknown)',
        is_sysadmin:  !!row.is_sysadmin,
        granted_at:   row.granted_at
            ? new Date(row.granted_at * 1000).toISOString().slice(0, 10)
            : '',
    });
}

// --- Apply user filter ---
let eventList = Object.values(events);
if (userFilter) {
    eventList = eventList.filter(e =>
        e.permissions.some(p =>
            p.email?.includes(userFilter) || p.user_id?.includes(userFilter)
        )
    );
}

// --- Attach camporee metadata ---
for (const e of eventList) {
    Object.assign(e, readCamporeeTitle(e.event_id));
    const dirPath = path.join(WORKSPACE_PATH, e.event_id);
    e.dir_exists = fs.existsSync(dirPath);
}

// --- Output ---
if (jsonMode) {
    const output = {
        workspace_path: WORKSPACE_PATH,
        generated_at:   new Date().toISOString(),
        events:         eventList,
        orphan_dirs:    orphanDirs,
    };
    console.log(JSON.stringify(output, null, 2));
    process.exit(0);
}

// --- Table output ---
const COLS = {
    event_id:    12,
    title:       32,
    year:         6,
    role:         8,
    email:       32,
    display_name: 20,
    granted_at:  12,
    dir:          5,
};

function pad(str, len) {
    const s = String(str ?? '');
    return s.length > len ? s.slice(0, len - 1) + '…' : s.padEnd(len);
}

const divider = Object.values(COLS).map(n => '─'.repeat(n)).join('─┼─');
const header  = [
    pad('event_id',    COLS.event_id),
    pad('title',       COLS.title),
    pad('year',        COLS.year),
    pad('role',        COLS.role),
    pad('email',       COLS.email),
    pad('display_name',COLS.display_name),
    pad('granted_at',  COLS.granted_at),
    pad('dir?',        COLS.dir),
].join(' │ ');

console.log(`\nWorkspace audit — ${new Date().toISOString()}`);
console.log(`DB:         ${process.env.CONDUCTOR_DB_PATH || '(default)'}`);
console.log(`Workspaces: ${WORKSPACE_PATH}`);
if (userFilter) console.log(`Filter:     ${userFilter}`);
console.log(`Events:     ${eventList.length}  |  Orphan dirs: ${orphanDirs.length}\n`);

console.log(header);
console.log(divider);

for (const e of eventList) {
    for (const p of e.permissions) {
        console.log([
            pad(e.event_id,    COLS.event_id),
            pad(e.title,       COLS.title),
            pad(e.year,        COLS.year),
            pad(p.role,        COLS.role),
            pad(p.email,       COLS.email),
            pad(p.display_name,COLS.display_name),
            pad(p.granted_at,  COLS.granted_at),
            pad(e.dir_exists ? '✓' : '✗', COLS.dir),
        ].join(' │ '));
    }
}

if (orphanDirs.length) {
    console.log('\n── Orphan workspace dirs (no DB entry) ──');
    for (const d of orphanDirs) {
        const meta = readCamporeeTitle(d);
        console.log(`  ${d}  →  ${meta.title} ${meta.year}`);
    }
}

console.log('');
